File: db4o.net/Db4objects.Db4o.Tests/Db4objects.Db4o.Tests/Common/Staging/ActivatorIsNotStoredTestCase.cs
/* Copyright (C) 2004 - 2011  Versant Inc.  http://www.db4o.com */

using System;
using Db4oUnit;
using Db4oUnit.Extensions;
using Db4objects.Db4o.Activation;
using Db4objects.Db4o.Config;
using Db4objects.Db4o.Query;
using Db4objects.Db4o.TA;
using Db4objects.Db4o.Tests.Common.Staging;

namespace Db4objects.Db4o.Tests.Common.Staging
{
	/// <summary>
	/// COR-1003
	/// If the Activator is not declared as transient,
	/// it will get restored as null from the database
	/// when activation happens.
	/// </summary>
	/// <remarks>
	/// COR-1003
	/// If the Activator is not declared as transient,
	/// it will get restored as null from the database
	/// when activation happens.
	/// </remarks>
	public class ActivatorIsNotStoredTestCase : AbstractDb4oTestCase
	{
		public class Item : IActivatable
		{
			public string _name;

			public IActivator _activator;

			public virtual void Activate(ActivationPurpose purpose)
			{
				if (this._activator != null)
				{
					this._activator.Activate(purpose);
				}
			}

			public virtual void Bind(IActivator activator)
			{
				if (this._activator == activator)
				{
					return;
				}
				if (activator != null && this._activator != null)
				{
					throw new InvalidOperationException();
				}
				this._activator = activator;
			}

			internal Item(ActivatorIsNotStoredTestCase _enclosing)
			{
				this._enclosing = _enclosing;
			}

			private readonly ActivatorIsNotStoredTestCase _enclosing;
		}

		/// <exception cref="System.Exception"></exception>
		protected override void Configure(IConfiguration config)
		{
			config.Add(new TransparentActivationSupport());
			config.Add(new TransparentActivationSupport());
		}

		/// <exception cref="System.Exception"></exception>
		protected override void Store()
		{
			ActivatorIsNotStoredTestCase.Item item = new ActivatorIsNotStoredTestCase.Item(this
				);
			item._name = "one";
			Store(item);
		}

		public virtual void Test()
		{
			ActivatorIsNotStoredTestCase.Item item = ((ActivatorIsNotStoredTestCase.Item)RetrieveOnlyInstance
				(typeof(ActivatorIsNotStoredTestCase.Item)));
			item.Activate(ActivationPurpose.Write);
			Assert.IsNotNull(item._activator);
			Store(item);
			IQuery q = NewQuery(typeof(IActivator));
			Assert.AreEqual(0, q.Execute().Count);
		}
	}
}

File: db4oj.tests/src/com/db4o/db4ounit/common/staging/ActivatorIsNotStoredTestCase.java
/* Copyright (C) 2009  Versant Inc.  http://www.db4o.com */

package com.db4o.db4ounit.common.staging;

import com.db4o.activation.*;
import com.db4o.config.*;
import com.db4o.query.*;
import com.db4o.ta.*;

import db4ounit.*;
import db4ounit.extensions.*;

/**
 * COR-1003
 * If the Activator is not declared as transient,
 * it will get restored as null from the database
 * when activation happens.
 */
public class ActivatorIsNotStoredTestCase extends AbstractDb4oTestCase{
	
	public class Item implements Activatable {
		
		public String _name;
		
		public Activator _activator;

		public void activate(ActivationPurpose purpose) {
			if(_activator != null) {
				_activator.activate(purpose);
			}
		}

		public void bind(Activator activator) {
	    	if (_activator == activator) {
	    		return;
	    	}
	    	if (activator != null && _activator != null) {
	            throw new IllegalStateException();
	        }
			_activator = activator;
		}
	}
	
	@Override
	protected void configure(Configuration config) throws Exception {
		config.add(new TransparentActivationSupport());
		config.add(new TransparentActivationSupport());
	}
	
	@Override
	protected void store() throws Exception {
		Item item = new Item();
		item._name = "one";
		store(item);
	}
	
	public void test(){
		Item item = retrieveOnlyInstance(Item.class);
		item.activate(ActivationPurpose.WRITE);
		Assert.isNotNull(item._activator);
		
		store(item);
		Query q = newQuery(Activator.class);
		Assert.areEqual(0, q.execute().size());
	}

}
